stone/frontend/ir_generator.py

```diff
@@ -1248,3 +1259,235 @@ def mk_route_schema():
+
+    def _find_dependencies(self, data_types):
+        output_types = defaultdict(list)
+        output_routes = defaultdict(set)
+        seen = set()
+        for t in data_types:
+            self._find_dependencies_recursive(t, seen, output_types, output_routes)
+        return output_types, output_routes
+
+    def _find_dependencies_recursive(self, data_type, seen, output_types,
+                                     output_routes, type_context=None):
+        # Define a function that recursively traverses data types and populates
+        # the data structures defined above.
+        if data_type in seen:
+            # if we've visited a data type already, no need to revisit
+            return
+        elif is_primitive_type(data_type):
+            # primitive types represent leaf nodes in the tree
+            return
+        elif is_struct_type(data_type) or is_union_type(data_type):
+            # recurse on fields and parent types for structs and unions
+            # also recurse on enumerated subtypes for structs if present
+            seen.add(data_type)
+            output_types[data_type.namespace.name].append(data_type)
+            for field in data_type.all_fields:
+                self._find_dependencies_recursive(field, seen, output_types, output_routes,
+                                                  type_context=data_type)
+            if data_type.parent_type is not None:
+                self._find_dependencies_recursive(data_type.parent_type, seen, output_types,
+                                                  output_routes)
+            if data_type.doc is not None:
+                doc_types, routes_by_ns = self._get_data_types_and_routes_from_doc_ref(
+                    data_type.doc, data_type.namespace.name)
+                for t in doc_types:
+                    self._find_dependencies_recursive(t, seen, output_types, output_routes)
+                for namespace_name, routes in routes_by_ns.items():
+                    route_namespace = self.api.namespaces[namespace_name]
+                    for route in routes:
+                        output_routes[namespace_name].add(route)
+                        route_types = route_namespace.get_route_io_data_types_for_route(route)
+                        for route_type in route_types:
+                            self._find_dependencies_recursive(route_type, seen, output_types,
+                                                              output_routes)
+            if is_struct_type(data_type) and data_type.has_enumerated_subtypes():
+                for subtype in data_type.get_enumerated_subtypes():
+                    self._find_dependencies_recursive(subtype, seen, output_types, output_routes,
+                                                      type_context=data_type)
+        elif is_alias(data_type) or is_field_type(data_type):
+            assert (is_field_type(data_type)) == (type_context is not None)
```

Comment:
this is a strange way to write this assertion case.  Am I correct in understanding that you want `is_field_type(data_type) XOR (type_context)